ext/storage_sources/azure_store/azure_connection.cpp

```diff
@@ -90,8 +90,26 @@ azure_connection::delete_object(const std::string &object_key) const
     return 0;
 }
 
+// Reads an object and outputs the content into a buffer.
 int
-azure_connection::get_object(const std::string &path) const
+azure_connection::read_object(
+  const std::string &object_key, off_t offset, size_t len, void *&buf) const
 {
+    auto blob_client = _azure_client.GetBlockBlobClient(_object_prefix + object_key);
+    auto blob_properties = blob_client.GetProperties().Value;
+    std::vector<uint8_t> downloaded_blob(blob_properties.BlobSize);
+
+    // Downloads the content of the blob to the vector downloaded_blob.
+    blob_client.DownloadTo(downloaded_blob.data(), downloaded_blob.size());
+    if (offset + len > downloaded_blob.size()) {
+        std::cerr << "Reading past end of file!" << std::endl;
+        return -1;
+    }
+
+    // Construct a string using the vector downloaded_blob's content between the offset and length
+    // supplied.
+    std::string content =
+      std::string(downloaded_blob.begin() + offset, downloaded_blob.begin() + offset + len);
+    buf = &content;
```

Comment:
I think the types here are a bit odd. The blob contents are just random binaries, right? If so, there's every chance they contain null bytes or other non-printable junk. I would consider truncating `downloaded_blob` and returning that, e.g. `downloaded_blob.erase(downloaded_blob.begin(), downloaded_blob.begin() + offset);` then `downloaded_blob.resize(len);`.

There will be a lifetime issue with returning the memory associated with the vector, I'm not sure it's possible to change the type of `buf` but I'd consider making it a `std::vector<uint8_t>&`. I think you have a lifetime issue with the current approach anyway - `content` owns the backing memory, and will release it when the function returns.

Also - how can the caller know the length of what gets returned in `buf`?

I think these questions go away if the blob contents are guaranteed to be a reasonably sensible string.

Reply:
right yeah I had a conversation with @jiechenbo about this to pass in the offset and length directly into download to so that we avoid having to deal with the whole converting to string and setting the start and end based on the buffer. In addition to that I have changed the function signature to take in a `std::vector<uint8_t> &`

Reply:
In regards to the void * this is needed so i've just typecasted the result of the downloadTo into the supplied buffer which is a void pointer. The changes can be seen here https://github.com/wiredtiger/wiredtiger/pull/8738/commits/7910412c1a928e07a480f4fb72bbf979155a4653